traffic_ops/traffic_ops_golang/riak.go

```diff
@@ -0,0 +1,267 @@
+	// build the fetch command
+	cmd, err := riak.NewFetchValueCommandBuilder().
+		WithBucket(bucket).
+		WithKey(key).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+	// create and start a riak cluster
+	cluster, err := getRiakCluster(db, cfg, 12)
+	if err != nil {
+		log.Errorf("%v\n", err)
+		return nil, err
+	}
+	defer func() {
+		if err := cluster.Stop(); err != nil {
+			log.Errorf("%v\n", err)
+		}
+	}()
+	if err = cluster.Start(); err != nil {
+		return nil, err
+	}
+	if err = cluster.Execute(cmd); err != nil {
+		return nil, err
+	}
+	fvc := cmd.(*riak.FetchValueCommand)
+
+	return fvc, err
+}
+
+// endpoint handler for fetching uri signing keys from riak
+func urisignkeysHandler(db *sqlx.DB, cfg Config) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleErr := func(err error, status int) {
+			log.Errorf("%v %v\n", r.RemoteAddr, err)
+			w.WriteHeader(status)
+			fmt.Fprintf(w, http.StatusText(status))
+		}
+
+		ctx := r.Context()
+		pathParams, err := getPathParams(ctx)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		xmlId := pathParams["xml-id"]
+
+		fvc, err := fetchObject(xmlId, cdn_uri_keys_bucket, db, cfg)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		resp, err := getStringValue(fvc.Response)
+		if err != nil {
+			handleErr(err, http.StatusNotFound)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		fmt.Fprintf(w, "%s", resp)
+	}
+}
+
+// returns a riak cluster of online riak nodes.
+func getRiakCluster(db *sqlx.DB, cfg Config, maxNodes int) (*riak.Cluster, error) {
+	riakServerQuery := `
+		SELECT s.host_name, s.domain_name FROM server s 
+		INNER JOIN type t on s.type = t.id 
+		INNER JOIN status st on s.status = st.id 
+		WHERE t.name = 'RIAK' AND st.name = 'ONLINE'
+		`
+
+	var nodes []*riak.Node
+	rows, err := db.Query(riakServerQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	for rows.Next() {
+		var s tc.Server
+		var n *riak.Node
+		if err := rows.Scan(&s.HostName, &s.DomainName); err != nil {
+			return nil, err
+		}
+		addr := fmt.Sprintf("%s.%s:%d", s.HostName, s.DomainName, RiakPort)
+		nodeOpts := &riak.NodeOptions{
+			RemoteAddress: addr,
+			AuthOptions:   cfg.RiakAuthOptions,
+		}
+		nodeOpts.AuthOptions.TlsConfig.ServerName = fmt.Sprintf("%s.%s", s.HostName, s.DomainName)
+		n, err := riak.NewNode(nodeOpts)
+		if err != nil {
+			return nil, err
+		}
+		nodes = append(nodes, n)
+	}
+
+	opts := &riak.ClusterOptions{
+		Nodes: nodes,
+	}
+	cluster, err := riak.NewCluster(opts)
+
+	return cluster, err
+}
+
+func validateURIKeyset(msg map[string]URISignerKeyset) error {
+	for key, value := range msg {
+		var found bool = false
+		issuer := key
+		renewalKid := value.RenewalKid
+		if issuer == "" {
+			return errors.New("JSON Keyset has no issuer")
+		}
+		if renewalKid == "" {
+			return errors.New("JSON Keyset has no renewal kid specified")
```

Comment:
Not all keysets require a renewal kid. Only precisely one does. If multiple keysets are provided for multiple issuers, only one may have a renewal kid.